src/Cache/ThrottlerCache.php

```diff
@@ -0,0 +1,146 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+namespace Sunspikes\Ratelimit\Cache;
+
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\Cache\CacheException as PsrCacheException;
+use Sunspikes\Ratelimit\Cache\Exception\CacheAdapterException;
+use Sunspikes\Ratelimit\Cache\Exception\ItemNotFoundException;
+
+class ThrottlerCache implements ThrottlerCacheInterface
+{
+    /**
+     * @var CacheItemPoolInterface
+     */
+    private $cacheItemPool;
+
+    /**
+     * ThrottlerCache constructor.
+     *
+     * @param CacheItemPoolInterface $cacheItemPool
+     */
+    public function __construct(CacheItemPoolInterface $cacheItemPool)
+    {
+        $this->cacheItemPool = $cacheItemPool;
+    }
+
+    /**
+     * @param string $key
+     *
+     * @return ThrottlerItemInterface
+     */
+    public function getItem(string $key): ThrottlerItemInterface
+    {
+        $item = $this->getThrottlerItem($key);
+
+        return $item;
+    }
+
+    /**
+     * @param string $key
+     *
+     * @return bool
+     */
+    public function hasItem(string $key): bool
+    {
+        return $this->cacheItemPool->hasItem($key);
+    }
+
+    /**
+     * @param string $key
+     * @param mixed  $item
+     * @return bool
+     */
+    public function setItem(string $key, ThrottlerItemInterface $item): bool
+    {
+        return $this->setThrottlerItem($key, $item);
+    }
+
+    /**
+     * @param string $key
+     */
+    public function removeItem(string $key)
+    {
+        $this->cacheItemPool->deleteItem($key);
+    }
+
+    /**
+     * @param string $key
+     *
+     * @return ThrottlerItemInterface
+     * @throws CacheAdapterException
+     * @throws ItemNotFoundException
+     */
+    private function getThrottlerItem(string $key): ThrottlerItemInterface
+    {
+        try {
+            $item = $this->cacheItemPool->getItem($key);
+
+            if ($item->isHit()) {
+                $params = json_decode($item->get(), true);
+                $class = $params['class'];
+
+                return $class::createFromArray($params['data']);
+            }
+        } catch (PsrCacheException $e) {
+        }
+
+        throw new ItemNotFoundException('Item not found.');
+    }
+
+    /**
+     * @param string                 $key
+     * @param ThrottlerItemInterface $item
+     *
+     * @return bool
+     * @throws CacheAdapterException
+     */
+    private function setThrottlerItem(string $key, ThrottlerItemInterface $item)
+    {
+        try {
+            $cacheItem = $this->cacheItemPool->getItem($key);
+
+            if ($cacheItem->isHit()) {
+                $this->cacheItemPool->deleteItem($key);
+            }
+        } catch (PsrCacheException $e) {
+            throw new CacheAdapterException($e->getMessage(), $e->getCode(), $e->getPrevious());
```

Comment:
Afaics all cache adapter operations (get, set, delete...) in case of critical errors will throw this one http://www.php-fig.org/psr/psr-6/#cacheexception